client/src/main/scala/skuber/api/package.scala

```diff
@@ -51,37 +52,115 @@ package object client {
-       result ++=")"
-       result.toString()
+   sealed trait AuthInfo
+
+   sealed trait AccessTokenAuth extends AuthInfo {
+     def accessToken: String
+   }
+
+   object NoAuth extends AuthInfo {
+     override def toString: String = "NoAuth"
+   }
+
+   final case class BasicAuth(userName: String, password: String) extends AuthInfo {
+     override def toString: String = s"${getClass.getSimpleName}(userName=$userName,password=<redacted>)"
+   }
+
+   final case class TokenAuth(token: String) extends AccessTokenAuth {
+
+     override def accessToken: String = token
+
+     override def toString: String = s"${getClass.getSimpleName}(token=<redacted>)"
+   }
+
+   final case class CertAuth(clientCertificate: PathOrData, clientKey: PathOrData, user: Option[String]) extends AuthInfo {
+     override def toString: String = StringBuilder.newBuilder
+       .append(getClass.getSimpleName)
+       .append("(")
+       .append {
+         clientCertificate match {
+           case Left(certPath: String) => "clientCertificate=" + certPath + " "
+           case Right(_) => "clientCertificate=<PEM masked> "
+         }
+       }
+       .append {
+         clientKey match {
+           case Left(certPath: String) => "clientKey=" + certPath + " "
+           case Right(_) => "clientKey=<PEM masked> "
+         }
+       }
+       .append("userName=")
+       .append(user.getOrElse(""))
+       .append(" )")
+       .mkString
+   }
+
+   sealed trait AuthProviderAuth extends AccessTokenAuth {
+     def name: String
+   }
+
+   // 'jwt' supports an oidc id token per https://kubernetes.io/docs/admin/authentication/#option-1---oidc-authenticator
+   // - but does not yet support token refresh
+   final case class OidcAuth(idToken: String) extends AuthProviderAuth {
+     override val name = "oidc"
+
+     override def accessToken: String = idToken
+
+     override def toString = """OidcAuth(idToken=<redacted>)"""
+   }
+
+   final case class GcpAuth private(private val config: GcpConfiguration) extends AuthProviderAuth {
+     override val name = "gcp"
+
+     @volatile private var refresh: GcpRefresh = new GcpRefresh(config.accessToken, config.expiry)
+
+     def refreshGcpToken(): GcpRefresh = {
+       val output = config.cmd.execute()
+       Json.parse(output).as[GcpRefresh]
      }
+
+     def accessToken: String = this.synchronized {
+       if(refresh.expired)
+         refresh = refreshGcpToken()
+
+       refresh.accessToken
+     }
+
+     override def toString =
+       """GcpAuth(accessToken=<redacted>)""".stripMargin
+   }
+
+   final private[client] case class GcpRefresh(accessToken: String, expiry: Instant) {
+     def expired: Boolean = Instant.now.isAfter(expiry.minusSeconds(20))
+   }
+
+  private[client] object GcpRefresh {
+    implicit val gcpRefreshReads: Reads[GcpRefresh] = (
+      (JsPath \ "credential" \ "access_token").read[String] and
+      (JsPath \ "credential" \ "token_expiry").read[Instant]
```

Comment:
This should not be hardcoded, but don't know how to use JsPath to apply what there is in the config, I should evaluate the JsPath in these

```
        expiry-key: '{.credential.token_expiry}'
        token-key: '{.credential.access_token}'
```